### ISSUE TITLE
Squiz.Strings.DoubleQuoteUsage broken for some escape codes

### ISSUE THREAD

**Issue by ElvenSpellmaker** [squizlabs/PHP_CodeSniffer#867]
This sniff is broken for `\e`, `\1` (etc.), and for `\u` all of which are valid uses for double quotes.

Also, your `composer.json` states a minimum PHP of `5.1.2` and then defines `\v` and `\f` as allowed characters which is not the case until PHP `5.2.5`. I'm not sure how one can test that sort of thing using your current testing framework.

An example:
```
----------------------------------------------------------------------
 33 | ERROR   | [x] String "\e[0;32mPASS\e[0m" does not require
    |         |     double quotes; use single quotes instead
----------------------------------------------------------------------
```
and
```
----------------------------------------------------------------------
 3 | ERROR | [x] String "\1" does not require double quotes; use
   |       |     single quotes instead
----------------------------------------------------------------------
```

Btw., `\e` has been available since `5.4.4`, and `\u` since `7.0.0`.

Link to the manual: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.double

Probably changing the `$allowedChars` array to this:
```php
        $allowedChars = array(
                         '\0',
                         '\1',
                         '\2',
                         '\3',
                         '\4',
                         '\5',
                         '\6',
                         '\7',
                         '\n',
                         '\r',
                         '\t',
                         '\x',
                         '\b',
                         '\'',
                        );

        if (phpversion() >= '5.2.5') {
            $allowedChars[] = '\v';
            $allowedChars[] = '\f';
        }

        if (phpversion() >= '5.4.4') {
            $allowedChars[] = '\e';
        }

        if (phpversion() >= '7.0.0') {
            $allowedChars[] = '\u';
        }
```
is best, but then how you go about testing that on different platforms with the current test framework is as said, impossible.

**Comment by gsherwood** [comment]
> is best, but then how you go about testing that on different platforms with the current test framework is as said, impossible.

You can report different errors for different PHP versions pretty easily, but you can't provide different fixed versions. But that ultimately doesn't matter as I wouldn't put PHP version exceptions in for these characters. It's better to always test for all of them because you can't assume that the version of PHP used for PHPCS is the same version that the code is designed to run on.

> Also, your composer.json states a minimum PHP of 5.1.2 and then defines \v and \f as allowed characters which is not the case until PHP 5.2.5

5.1.2 is what PHPCS attempts to run on. It doesn't mean the sniffs wont be checking for PHP features in newer versions. The tokenizer even replicates newer tokens so that you can run PHPCS on older PHP versions but still check code designed for newer ones.

But those issues don't really matter as you've provided more than enough information for me to fix this issue. So thanks for the report.

**Comment by gsherwood** [comment]
I've committed a fix. Thanks again.

**Comment by aik099** [comment]
@gsherwood , have you pushed that commit to GitHub? I don't see neither link to commit from this issue nor new commit in commit list on `master` branch.

**Comment by gsherwood** [comment]
> have you pushed that commit to GitHub?

I have now :) Thanks for noticing.

**Comment by aik099** [comment]
You're welcome :wink:
